cocos/3d/CCTerrain.cpp

```diff
@@ -1539,4 +1577,77 @@ Terrain::DetailMap::DetailMap()
     _detailMapSize = 35;
 }
 
+Terrain::Triangle::Triangle(Vec3 p1, Vec3 p2, Vec3 p3)
+{
+    _p1 = p1;
+    _p2 = p2;
+    _p3 = p3;
+}
+
+void Terrain::Triangle::transform(cocos2d::Mat4 matrix)
+{
+    matrix.transformPoint(&_p1);
+    matrix.transformPoint(&_p2);
+    matrix.transformPoint(&_p3);
+}
+
+bool Terrain::Triangle::getInsterctPoint(const Ray &ray, Vec3& interScetPoint)const
+{
```

Comment:
After several hours, I found the original link for the algorithm:
https://en.wikipedia.org/wiki/M%C3%B6ller%E2%80%93Trumbore_intersection_algorithm